File: flutter/lib/core/network/websocket_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

import '../constants/api_constants.dart';

enum WebSocketStatus { disconnected, connecting, connected, error }

class WebSocketClient {
  WebSocketChannel? _channel;
  final _statusController = StreamController<WebSocketStatus>.broadcast();
  final _messageController = StreamController<Map<String, dynamic>>.broadcast();

  WebSocketStatus _status = WebSocketStatus.disconnected;
  String? _authToken;
  String? _deviceId;
  Timer? _reconnectTimer;

  Stream<WebSocketStatus> get statusStream => _statusController.stream;
  Stream<Map<String, dynamic>> get messageStream => _messageController.stream;
  WebSocketStatus get status => _status;

  void setCredentials(String authToken, String deviceId) {
    _authToken = authToken;
    _deviceId = deviceId;
  }

  void connect() {
    if (_authToken == null || _deviceId == null) {
      return;
    }

    _updateStatus(WebSocketStatus.connecting);

    try {
      final wsUrl = Uri.parse(
        '${ApiConstants.baseUrl.replaceFirst('http', 'ws')}${ApiConstants.wsSync}',
      ).replace(
        queryParameters: {
          'token': _authToken,
          'device_id': _deviceId,
        },
      );

      _channel = WebSocketChannel.connect(wsUrl);
      _updateStatus(WebSocketStatus.connected);

      _channel!.stream.listen(
        (message) {
          try {
            final data = jsonDecode(message as String) as Map<String, dynamic>;
            _messageController.add(data);
          } catch (e) {
            // Ignore invalid messages
          }
        },
        onError: (_) {
          _updateStatus(WebSocketStatus.error);
          _scheduleReconnect();
        },
        onDone: () {
          _updateStatus(WebSocketStatus.disconnected);
          _scheduleReconnect();
        },
      );
    } catch (e) {
      _updateStatus(WebSocketStatus.error);
      _scheduleReconnect();
    }
  }

  void disconnect() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _channel?.sink.close();
    _channel = null;
    _updateStatus(WebSocketStatus.disconnected);
  }

  void send(Map<String, dynamic> message) {
    if (_status == WebSocketStatus.connected && _channel != null) {
      _channel!.sink.add(jsonEncode(message));
    }
  }

  void _updateStatus(WebSocketStatus status) {
    _status = status;
    _statusController.add(status);
  }

  void _scheduleReconnect() {
    if (_reconnectTimer != null) return;

    _reconnectTimer = Timer(const Duration(seconds: 5), () {
      _reconnectTimer = null;
      if (_status != WebSocketStatus.connected) {
        connect();
      }
    });
  }

  void dispose() {
    disconnect();
    _statusController.close();
    _messageController.close();
  }
}

File: web/src/sync/websocket-client.ts
import ReconnectingWebSocket from 'reconnecting-websocket';
import { Subject, BehaviorSubject } from 'rxjs';
import { API_CONSTANTS } from '../core/constants/api.constants';

export type WebSocketStatus = 'disconnected' | 'connecting' | 'connected' | 'error';

export class WebSocketClient {
  private ws: ReconnectingWebSocket | null = null;
  private statusSubject = new BehaviorSubject<WebSocketStatus>('disconnected');
  private messageSubject = new Subject<unknown>();

  public status$ = this.statusSubject.asObservable();
  public message$ = this.messageSubject.asObservable();

  constructor(
    private authToken: string,
    private deviceId: string
  ) {}

  connect(): void {
    if (this.ws) return;

    this.updateStatus('connecting');

    const wsUrl = new URL(
      `${API_CONSTANTS.baseUrl.replace('http', 'ws')}${API_CONSTANTS.wsSync}`
    );
    wsUrl.searchParams.set('token', this.authToken);
    wsUrl.searchParams.set('device_id', this.deviceId);

    this.ws = new ReconnectingWebSocket(wsUrl.toString());

    this.ws.onopen = () => {
      this.updateStatus('connected');
    };

    this.ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        this.messageSubject.next(data);
      } catch (e) {
        console.error('WebSocket message parse error:', e);
      }
    };

    this.ws.onerror = () => {
      this.updateStatus('error');
    };

    this.ws.onclose = () => {
      this.updateStatus('disconnected');
    };
  }

  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  send(message: unknown): void {
    if (this.ws && this.statusSubject.value === 'connected') {
      this.ws.send(JSON.stringify(message));
    }
  }

  private updateStatus(status: WebSocketStatus): void {
    this.statusSubject.next(status);
  }

  dispose(): void {
    this.disconnect();
    this.statusSubject.complete();
    this.messageSubject.complete();
  }
}
